src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java

```diff
@@ -83,36 +94,82 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-    String requestedInterviewerId = request.getParameter("interviewer");
-    String requestedIntervieweeId = request.getParameter("interviewee");
-    String userEmail = UserServiceFactory.getUserService().getCurrentUser().getEmail();
-    String userId = UserServiceFactory.getUserService().getCurrentUser().getUserId();
-    // Since UserId does not have a valid Mock, if the id is null (as when testing), it will be
-    // replaced with this hashcode.
-    if (userId == null) {
-      userId = String.format("%d", userEmail.hashCode());
+    String intervieweeEmail = userService.getCurrentUser().getEmail();
+    String intervieweeId = getUserId();
+
+    InterviewPostRequest postRequest;
+    try {
+      postRequest = new Gson().fromJson(getJsonString(request), InterviewPostRequest.class);
+    } catch (Exception JsonSyntaxException) {
+      response.sendError(400);
+      return;
     }
-    // The default key for a scheduledInterview being stored in datastore
-    long defaultKey = -1;
-    if ((!requestedInterviewerId.equals(userId)) && (!requestedIntervieweeId.equals(userId))) {
-      response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+    if (!postRequest.allFieldsPopulated()) {
+      response.sendError(400);
       return;
     }
+
+    String interviewerCompany = postRequest.getCompany();
+    String interviewerJob = postRequest.getJob();
+    String utcStartTime = postRequest.getUtcStartTime();
+
+    TimeRange range;
+
     try {
-      ScheduledInterview scheduledInterview =
-          ScheduledInterview.create(
-              defaultKey,
-              new TimeRange(
-                  Instant.parse(request.getParameter("startTime")),
-                  Instant.parse(request.getParameter("endTime"))),
-              requestedInterviewerId,
-              requestedIntervieweeId);
-      scheduledInterviewDao.create(scheduledInterview);
-      return;
+      range =
+          new TimeRange(
+              Instant.parse(utcStartTime), Instant.parse(utcStartTime).plus(1, ChronoUnit.HOURS));
     } catch (DateTimeParseException e) {
       response.sendError(400, e.getMessage());
       return;
     }
+
+    List<Availability> availabilitiesInRange =
+        availabilityDao.getInRangeForAll(range.start(), range.end());
+    List<Person> allAvailableInterviewers =
+        ShowInterviewersServlet.getPossiblePeople(
+            personDao, availabilityDao, availabilitiesInRange, range);
+    List<String> possibleInterviewers =
+        getPossibleInterviewerIds(allAvailableInterviewers, interviewerCompany, interviewerJob);
+
+    int randomNumber = (int) (Math.random() * possibleInterviewers.size());
+    String interviewerId = possibleInterviewers.get(randomNumber);
+
+    scheduledInterviewDao.create(
+        ScheduledInterview.create(-1, range, interviewerId, intervieweeId));
+
+    // Since an interview was scheduled, both parties' availabilities must be updated
```

Comment:
This is not something you're going to be able to fix with the time remaining, but I think this represents a possible race condition.

Basically, you create a could scheduled interview for two interviewees and the same interviewer at the same time. Both ScheduledInterview would get created, but only one interview can actually impact the availability.